core/finality-grandpa/src/until_imported.rs

```diff
@@ -0,0 +1,565 @@
+//!
+//! This is used for votes and commit messages currently.
+
+use super::{BlockStatus, Error, SignedMessage, CompactCommit};
+
+use client::ImportNotifications;
+use futures::prelude::*;
+use futures::stream::Fuse;
+use parking_lot::Mutex;
+use runtime_primitives::traits::{Block as BlockT, Header as HeaderT, NumberFor};
+use substrate_primitives::AuthorityId;
+use tokio::timer::Interval;
+
+use std::collections::{HashMap, VecDeque};
+use std::sync::{atomic::{AtomicUsize, Ordering}, Arc};
+use std::time::{Duration, Instant};
+
+// something which will block until imported.
+pub(crate) trait BlockUntilImported<Block: BlockT>: Sized {
+	// the type that is blocked on.
+	type Blocked;
+
+	/// new incoming item. For all internal items,
+	/// check if they require to be waited for.
+	/// if so, call the `Wait` closure.
+	/// if they are ready, call the `Ready` closure.
+	fn schedule_wait<S, Wait, Ready>(
+		input: Self::Blocked,
+		status_check: &S,
+		wait: Wait,
+		ready: Ready,
+	) -> Result<(), Error> where
+		S: BlockStatus<Block>,
+		Wait: FnMut(Block::Hash, Self),
+		Ready: FnMut(Self::Blocked);
+
+	/// called when the wait has completed. The canonical number is passed through
+	/// for further checks.
+	fn wait_completed(self, canon_number: NumberFor<Block>) -> Option<Self::Blocked>;
+}
+
+/// Buffering imported messages until blocks with given hashes are imported.
+pub(crate) struct UntilImported<Block: BlockT, Status, I, M: BlockUntilImported<Block>> {
+	import_notifications: Fuse<ImportNotifications<Block>>,
+	status_check: Status,
+	inner: Fuse<I>,
+	ready: VecDeque<M::Blocked>,
+	check_pending: Interval,
+	pending: HashMap<Block::Hash, Vec<M>>,
+}
+
+impl<Block: BlockT, Status, I: Stream, M> UntilImported<Block, Status, I, M>
+	where Status: BlockStatus<Block>, M: BlockUntilImported<Block>
+{
+	/// Create a new `UntilImported` wrapper.
+	pub(crate) fn new(
+		import_notifications: ImportNotifications<Block>,
+		status_check: Status,
+		stream: I,
+	) -> Self {
+		// how often to check if pending messages that are waiting for blocks to be
+		// imported can be checked.
+		//
+		// the import notifications interval takes care of most of this; this is
+		// used in the event of missed import notifications
+		const CHECK_PENDING_INTERVAL: Duration = Duration::from_secs(5);
+		let now = Instant::now();
+
+		let check_pending = Interval::new(now + CHECK_PENDING_INTERVAL, CHECK_PENDING_INTERVAL);
+		UntilImported {
+			import_notifications: import_notifications.fuse(),
+			status_check,
+			inner: stream.fuse(),
+			ready: VecDeque::new(),
+			check_pending,
+			pending: HashMap::new(),
+		}
+	}
+}
+
+impl<Block: BlockT, Status, I, M> Stream for UntilImported<Block, Status, I, M> where
+	Status: BlockStatus<Block>,
+	I: Stream<Item=M::Blocked,Error=Error>,
+	M: BlockUntilImported<Block>,
+{
+	type Item = M::Blocked;
+	type Error = Error;
+
+	fn poll(&mut self) -> Poll<Option<M::Blocked>, Error> {
+		loop {
+			match self.inner.poll() {
```

Comment:
If you write `self.inner.poll()?` you can remove all the `Ok(..)` and the `Err(..)`.